esp/services/ws_fs/ws_fsService.cpp

```diff
@@ -1916,38 +1960,95 @@ void CFileSprayEx::readAndCheckSpraySourceReq(MemoryBuffer& srcxml, const char*
         if (containsRelPaths(sourcePath)) //Detect a path like: a/../../../f
             throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Invalid path %s", sourcePath.str());
 
-        if (!isEmptyString(srcPlane))
+        getStandardPosixPath(sourcePathReq, sourcePath.str());
+        StringArray files;
+        files.appendList(sourcePathReq, ","); // handles comma separated files
+
+        Owned<IPropertyTree> dropZone;
+        sourceIPReq.set(srcIP).trim();
+        sourcePlaneReq.set(srcPlane).trim();
+        if (!sourcePlaneReq.isEmpty())
         {
-            Owned<IPropertyTree> dropZone = getDropZonePlane(srcPlane);
+            dropZone.setown(getDropZonePlane(sourcePlaneReq));
             if (!dropZone)
-                throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Unknown landing zone: %s", srcPlane);
-            const char * dropZonePlanePath = dropZone->queryProp("@prefix");
-            if (isAbsolutePath(sourcePath))
+                throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Unknown landing zone: %s", sourcePlaneReq.str());
+            getDropZoneHost(sourcePlaneReq, dropZone, sourceIPReq);
+        }
+        else
+        {
+            if (sourceIPReq.isEmpty())
+                throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Source network IP not specified.");
+            StringBuffer path;
+            ForEachItemIn(i, files)
             {
-                if (!startsWith(sourcePath, dropZonePlanePath))
-                    throw makeStringException(ECLWATCH_INVALID_INPUT, "Invalid source path");
+                const char* file = files.item(i);
+                if (!isEmptyString(file))
+                {
+                    path.set(file);
+                    break;
+                }
             }
-            else
+            if (!path.isEmpty())
             {
-                StringBuffer s(sourcePath);
-                sourcePath.set(dropZonePlanePath);
-                addNonEmptyPathSepChar(sourcePath);
-                sourcePath.append(s);
+                dropZone.setown(getAndValidateDropZone(path, sourceIPReq));
+                if (dropZone)
+                    sourcePlaneReq.append(dropZone->queryProp("@name"));
             }
-            getDropZoneHost(srcPlane, dropZone, sourceIPReq);
-            sourcePlaneReq.append(srcPlane);
         }
-        else
+        if (sourcePlaneReq.isEmpty()) // must be true, unless bare-metal and isDropZoneRestrictionEnabled()==false
         {
-            sourceIPReq.set(srcIP).trim();
-            if (sourceIPReq.isEmpty())
-                throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Source network IP not specified.");
-            Owned<IPropertyTree> plane = getAndValidateDropZone(sourcePath, sourceIPReq);
-            if (plane)
-                sourcePlaneReq.append(plane->queryProp("@name"));
+            validateHostReqInSprayPathReq(sourceIPReq, sourcePathReq); //The sourceIPReq is from the original request.
```

Comment:
related to comment on line 1987, validateHostReqInSprayPathReq will be iterate through files again.

There are 3 separate coded iterations through 'files' in this function (1 in validateHostReqInSprayPathReq ), it feels like a single loop that dealt with dropzone, non-dropzone and UNC paths would be clearer .

Reply:
In https://github.com/hpcc-systems/HPCC-Platform/pull/17356/commits/e51b2e9e539c3589cd2f4099ee2422df548b1664, I tried doing them in a single loop. But, it was not easy to read (understand the logic). I think the 3 loops here are better for read. For one spray, only 2 of them are called for either a plane is found or not found. In both cases, the loop around line 1987 will be called and the loop should not cost much.

Reply:
ok, well as long as all paths have been tested, this code will be refactored soon anyway.